packages/core-utils/src/fees.ts

```diff
@@ -0,0 +1,95 @@
+/**
+ * Fee related serialization and deserialization
+ */
+
+import { BigNumber } from 'ethers'
+import { remove0x } from './common'
+
+const hundredMillion = BigNumber.from(100_000_000)
+const txDataZeroGas = 4
+const txDataNonZeroGasEIP2028 = 16
+const overhead = 4200
+
+function encode(
+  data: Buffer | string,
+  l1GasPrice: BigNumber,
+  l2GasLimit: BigNumber,
+  l2GasPrice: BigNumber
+): BigNumber {
+  const l1GasLimit = calculateL1GasLimit(data)
```

Comment:
This should throw an error if the values do not satisfy the correct equations, see https://github.com/ethereum-optimism/optimism/issues/823#issuecomment-844557937